source/s3_auto_ranged_put.c

```diff
@@ -1323,22 +1323,76 @@ static struct aws_future_http_message *s_s3_prepare_complete_multipart_upload(st
 static void s_s3_prepare_complete_multipart_upload_on_skipping_done(void *user_data) {
 
     struct aws_s3_prepare_complete_multipart_upload_job *complete_mpu_prep = user_data;
-    struct aws_s3_request *request = complete_mpu_prep->request;
-    struct aws_s3_meta_request *meta_request = request->meta_request;
 
     int error_code = aws_future_void_get_error(complete_mpu_prep->asyncstep_skip_remaining_parts);
     if (error_code != AWS_ERROR_SUCCESS) {
         s_s3_prepare_complete_multipart_upload_finish(complete_mpu_prep, error_code);
         return;
     }
 
-    aws_byte_buf_init(
-        &request->request_body, meta_request->allocator, s_complete_multipart_upload_init_body_size_bytes);
-
     /* Async steps complete, finish up job */
     s_s3_prepare_complete_multipart_upload_finish(complete_mpu_prep, AWS_ERROR_SUCCESS);
 }
 
+/* Allow user to review what we've uploaded, and fail the meta-request if they don't approve. */
+static int s_s3_review_multipart_upload(struct aws_s3_request *request) {
+    struct aws_s3_meta_request *meta_request = request->meta_request;
+    struct aws_s3_auto_ranged_put *auto_ranged_put = meta_request->impl;
+
+    /* If user registered no callback, then success! */
+    if (meta_request->upload_review_callback == NULL) {
+        return AWS_OP_SUCCESS;
+    }
+
+    /* Prepare review info */
+    struct aws_s3_upload_review review = {
+        .checksum_algorithm = meta_request->checksum_config.checksum_algorithm,
+    };
+
+    /* BEGIN CRITICAL SECTION */
+    aws_s3_meta_request_lock_synced_data(meta_request);
+
+    review.part_count = aws_array_list_length(&auto_ranged_put->synced_data.part_list);
+
+    if (review.part_count > 0) {
+        review.part_array =
+            aws_mem_calloc(meta_request->allocator, review.part_count, sizeof(struct aws_s3_upload_part_review));
+
+        for (size_t part_index = 0; part_index < review.part_count; ++part_index) {
+            struct aws_s3_mpu_part_info *part;
+            aws_array_list_get_at(&auto_ranged_put->synced_data.part_list, &part, part_index);
+
+            struct aws_s3_upload_part_review *part_review = &review.part_array[part_index];
+            part_review->size = part->size;
+            part_review->checksum = aws_byte_cursor_from_buf(&part->checksum_base64);
+        }
+    }
+
+    aws_s3_meta_request_unlock_synced_data(meta_request);
+    /* END CRITICAL SECTION */
+
+    /* Invoke callback */
+    int error_code = AWS_ERROR_SUCCESS;
+    if (meta_request->upload_review_callback(meta_request, &review, meta_request->user_data) != AWS_OP_SUCCESS) {
+        error_code = aws_last_error_or_unknown();
+    }
```

Comment:
Are we expecting the user to set a proper error code? I think we should at least replace it with a relevant error in the event of an unknown case. Also, we should mention in the documentation of the callback how to raise the error.

Reply:
I had the same thought when I wrote this
Users ARE expected to properly raise errors.
The documentation is bad because we didn't want to add 10 lines explaining proper error handling to every single callback.

I went back and forth on what to do... but ended up adding a short 3 line explanation to every callback in s3_client.h

Reply:
user's aren't just expected to raise errors properly, it's required
otherwise they'll end up with a random error code
if no other error has EVER been raised on this thread, they'll end up with AWS_ERROR_UNKNOWN
but if any other error has EVER been raised on this thread, that's the error code they'll end up seeing

IMHO: the whole aws_raise_error() system is very very fragile